test/integration/tcp_proxy_integration_test.cc

```diff
@@ -574,14 +534,150 @@ TEST_P(TcpProxyIntegrationTest, AccessLog) {
+                                ->mutable_lb_endpoints(0);
+        lb_endpoint->mutable_endpoint()
+            ->mutable_address()
+            ->mutable_socket_address()
+            ->set_port_value(1);
+      });
+  config_helper_.skipPortUsageValidation();
+  enableHalfClose(false);
+  initialize();
+  auto tcp_client = makeTcpConnection(lookupPort("tcp_proxy"));
+  ASSERT_TRUE(tcp_client->write("bonjour"));
+  tcp_client->waitForDisconnect();
+
+  tcp_client.reset();
+  test_server_.reset();
+  auto log_result = waitForAccessLog(listener_access_log_name_);
+  EXPECT_THAT(log_result,
+              MatchesRegex(fmt::format("DOWNSTREAM_WIRE_BYTES_SENT=0 "
+                                       "DOWNSTREAM_WIRE_BYTES_RECEIVED=0 "
+                                       "UPSTREAM_WIRE_BYTES_SENT=0 "
+                                       "UPSTREAM_WIRE_BYTES_RECEIVED=0"
+                                       "\r?\n.*")));
+}
+
+// Make sure no bytes are logged when no data is sent.
+TEST_P(TcpProxyIntegrationTest, TcpProxyNoDataBytesMeter) {
+  setupByteMeterAccessLog(this);
+  initialize();
+
+  FakeRawConnectionPtr fake_upstream_connection;
+  auto tcp_client = makeTcpConnection(lookupPort("tcp_proxy"));
+  ASSERT_TRUE(
+      fake_upstreams_[0]->waitForRawConnection(fake_upstream_connection));
+  tcp_client->close();
+  ASSERT_TRUE(fake_upstream_connection->close());
+
+  tcp_client.reset();
+  test_server_.reset();
+  auto log_result = waitForAccessLog(listener_access_log_name_);
+  EXPECT_THAT(log_result,
+              MatchesRegex(fmt::format("DOWNSTREAM_WIRE_BYTES_SENT=0 "
+                                       "DOWNSTREAM_WIRE_BYTES_RECEIVED=0 "
+                                       "UPSTREAM_WIRE_BYTES_SENT=0 "
+                                       "UPSTREAM_WIRE_BYTES_RECEIVED=0"
+                                       "\r?\n.*")));
+}
+
+// Test Byte Metering across multiple upstream/downstream connections
+TEST_P(TcpProxyIntegrationTest, TcpProxyBidirectionalBytesMeter) {
+  setupByteMeterAccessLog(this);
+  int client_count = 3;
+  int upstream_count = 3;
+
+  enableHalfClose(false);
+  config_helper_.addConfigModifier(
+      [this,
+       &upstream_count](envoy::config::bootstrap::v3::Bootstrap& bootstrap) {
+        auto* cluster =
+            bootstrap.mutable_static_resources()->mutable_clusters(0);
+        auto* load_assignment = cluster->mutable_load_assignment();
+        load_assignment->clear_endpoints();
+        for (int i = 0; i < upstream_count; ++i) {
+          auto locality = load_assignment->add_endpoints();
+          locality->add_lb_endpoints()->mutable_endpoint()->MergeFrom(
+              ConfigHelper::buildEndpoint(
+                  Network::Test::getLoopbackAddressString(version_)));
+        }
+      });
+
+  setUpstreamCount(upstream_count);
+  initialize();
+
+  u_long client_count_long = client_count;
+  u_long upstream_count_long = upstream_count;
```

Comment:
don't think we need these as longs, I think you should be fine doing `std::vector<IntegrationTCPClientPtr> clients(client_count)`